[JENKINS-53790] PodTemplate Container Docker image check

pyieh opened the pull request:
Check on ContainerWaitingState if the Docker image can't be found, prints message to build console and cancels and removes job from build queue.

[JENKINS-53790](https://issues.jenkins-ci.org/browse/JENKINS-53790)
Cleaned up and reforked, previous PR [#440](https://github.com/jenkinsci/kubernetes-plugin/pull/440#pullrequestreview-215310147)

Vlatombe commented on CHANGELOG.md:
```diff
@@ -11,6 +11,10 @@ Unreleased
 
+1.15.7
+------
+* Check on ContainerWaitingState if the Docker image can't be found, prints message to build console and cancels and removes job from build queue. [#497](https://github.com/jenkinsci/kubernetes-plugin/pull/497) [JENKINS-53790](https://issues.jenkins-ci.org/browse/JENKINS-53790)
```
Need to be moved to **Unreleased**

Vlatombe commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/AllContainersRunningPodWatcher.java:
```diff
@@ -72,7 +84,29 @@ boolean areAllContainersRunning(Pod pod) {
+                    String waitingStateMsg = waitingState.getMessage();
+                    if (waitingStateMsg != null && waitingStateMsg.contains("Back-off pulling image")) {
+                        runListener.error("Unable to pull Docker image \""+containerStatus.getImage()+"\". Check if image name is spelled correctly");
+                        Jenkins jenkins = Jenkins.getInstanceOrNull();
```
Using `Jenkins.get()` would simplify the code

Vlatombe commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/AllContainersRunningPodWatcher.java:
```diff
@@ -72,7 +84,29 @@ boolean areAllContainersRunning(Pod pod) {
+                            Queue q = jenkins.getQueue();
+                            for (Queue.Item item : q.getItems()) {
+                                Label itemLabel = item.getAssignedLabel();
+                                if (itemLabel != null && isCorrespondingLabels(itemLabel.getDisplayName(), pod.getMetadata().getName())) {
```
I understand the intent, but this is out of line here. Eventually could be added to exception handling in the launcher https://github.com/jenkinsci/kubernetes-plugin/blob/3225773a140eb590a4d2aaa8281e386812214f97/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java#L200-L208

pyieh replied:
Are you saying that the KubernetesLauncher should be in charge of canceling the job? We've seen that unless a job is canceled, it'll get put back in the queue and infinitely fail due to that bad pod template.

Vlatombe replied:
It would be a better place than in `AllContainersRunningPodWatcher`, certainly.

jglick commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/AllContainersRunningPodWatcher.java:
```diff
@@ -72,7 +84,29 @@ boolean areAllContainersRunning(Pod pod) {
+                                if (itemLabel != null && isCorrespondingLabels(itemLabel.getDisplayName(), pod.getMetadata().getName())) {
+                                    String itemTaskName = item.task.getFullDisplayName();
+                                    String jobName = getJobName(itemTaskName);
+                                    if (jobName.equals("")) {
```
```suggestion
                                    if (jobName.isEmpty()) {
```

jglick commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/AllContainersRunningPodWatcher.java:
```diff
@@ -72,7 +84,29 @@ boolean areAllContainersRunning(Pod pod) {
+                if (waitingState != null) {
+                    String waitingStateMsg = waitingState.getMessage();
+                    if (waitingStateMsg != null && waitingStateMsg.contains("Back-off pulling image")) {
+                        runListener.error("Unable to pull Docker image \""+containerStatus.getImage()+"\". Check if image name is spelled correctly");
```
This ought to be straightforward to add a test for (probably in `KubernetesPipelineTest`): a script requesting a bogus `image`, start building, verify that the build is aborted, `assertLogContains` this text (preferably factored into a `static` utility method for use from the test) and `org.jenkinsci.plugins.workflow.steps.durable_task.Messages.ExecutorStepExecution_queue_task_cancelled()`.

pyieh commented:
@jglick @Vlatombe please take a look at the new changes. 
We introduced a cleaner way of matching the created faulty pod to the build item in the queue that needs to be canceled. 
Also added new tests for this.

jglick commented on CHANGELOG.md:
```diff
@@ -9,6 +9,41 @@ See the full list of issues at [JIRA](https://issues.jenkins-ci.org/issues/?filt
+
+1.16.3
+------
+* **testing** Fixed test flakes [#496](https://github.com/jenkinsci/kubernetes-plugin/pull/496) [JENKINS-46076](https://issues.jenkins-ci.org/browse/JENKINS-46076)
```
Seems you misresolved at least one merge conflict, and there are still several outstanding merge conflicts. If you are not sure how to merge with upstream safely, better to trash the existing commits and force-push a fresh diff. (It is a good idea to check your own PR’s diff tab to verify that you are submitting what you think you are submitting.)

markyjackson-taulia commented:
@jglick @Vlatombe I had a chat with the person who opened this PR at JWDW yesterday. I explained what they needed to do in order to get this PR correct

Aghassi commented:
@markyjackson-taulia Would you be willing to elaborate here what the points were that are needed to make this PR acceptable? It may be something others can take up in a separate PR in the event that OP can't finish this one. This issue is very frustrating from an end user perspective when things fail due to typos/bad pulls etc. on our internal Jenkins system. If we could fix this, it would be a great usability benefit.

narayanan commented:
@jglick @Vlatombe Can you please elaborate what changes are needed to make this PR acceptable? 

In a managed Jenkins where user's do't get access to underlying kubernetes cluster, it will be very hard for them to debug any typo's in the defined podTemplate. If the underlying error is surfaced to the user in the build console, it will be very easy for them to identify the issue and fix the error.

jglick commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/AllContainersRunningPodWatcher.java:
```diff
@@ -68,7 +83,13 @@ boolean areAllContainersRunning(Pod pod) {
-                if (containerStatus.getState().getWaiting() != null) {
+                ContainerStateWaiting waitingState = containerStatus.getState().getWaiting();
+                if (waitingState != null) {
+//                    String waitingStateMsg = waitingState.getMessage();
```
If you are not changing the behavior of this method, better to revert all changes to it.

jglick commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:
```diff
@@ -124,12 +132,34 @@ public void launch(SlaveComputer computer, TaskListener listener) {
             try (Watch _w = client.pods().inNamespace(namespace1).withName(podName).watch(watcher)) {
                 watcher.await(template.getSlaveConnectTimeout(), TimeUnit.SECONDS);
             }
+//            } catch (IllegalStateException e) {
```
Just delete code if you think it is not useful.

jglick commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:
```diff
@@ -212,6 +242,17 @@ public void launch(SlaveComputer computer, TaskListener listener) {
 
+    /* itemTaskName is format of "part of <ORGANIZATION> <JOB NAME> >> <BRANCH> #<BUILD NUMBER> */
+    /* <ORGANIZATION> and <BRANCH> are only there if Pipeline created through BlueOcean, else just <JOB NAME>  */
+    private String getJobName(String itemTaskName) {
```
Now unused?

jglick commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java:
```diff
@@ -26,23 +36,6 @@
 import org.kohsuke.stapler.DataBoundConstructor;
 import org.kohsuke.stapler.DataBoundSetter;
 
-import com.google.common.base.Strings;
```
Set your editor to avoid reorganizing imports. Just creates diff churn and merge conflicts. (Ideally we would normalize all imports and enforce this at build time: strict lexicographic sort, no wildcards, no blank lines.)

jglick commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStepExecution.java:
```diff
@@ -113,9 +114,11 @@ public boolean start() throws Exception {
         newTemplate.setYamlMergeStrategy(step.getYamlMergeStrategy());
         if(run!=null) {
             newTemplate.getAnnotations().add(new PodAnnotation("buildUrl", ((KubernetesCloud)cloud).getJenkinsUrlOrDie()+run.getUrl()));
+            newTemplate.getAnnotations().add(new PodAnnotation("runUrl", run.getUrl()));
```
Unused?

jglick commented on src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesPipelineTest.java:
```diff
@@ -447,4 +448,25 @@ public void jenkinsSecretHidden() throws Exception {
         r.assertLogNotContains(jnlpMac, b);
     }
+
+//    @Test
```
Why commented out?

pyieh commented:
@jglick @Vlatombe please take a look at the changes. 

There seems to be a problem with the KubernetesPipelineTest.runInPodWithShowRawYamlFalse() test
```
Scripts not permitted to use staticMethod org.jenkinsci.plugins.workflow.cps.Safepoint safepoint. Administrators can decide whether to approve or reject this signature.
```
Which seems to be a settings problem? I don't run into this error when running tests locally (even running integration tests with `minikube`)

pyieh commented:
@jglick Looks like you made some changes and have fixed the issue I was seeing.

jglick commented:
> Looks like you made some changes

#655

pyieh commented:
@jglick @Vlatombe can I get a review for this? It was originally showing as all checks passing, but now checks are failing.

jglick commented:
Test failures seem to be infrastructural. Suggest `git pull origin master:master && git push` to trigger a new build.

pyieh commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/AllContainersRunningPodWatcher.java:
```diff
@@ -68,7 +77,13 @@ boolean areAllContainersRunning(Pod pod) {
+                    String waitingStateMsg = waitingState.getMessage();
+                    if (waitingStateMsg != null && waitingStateMsg.contains("Back-off pulling image")) {
+                        runListener.error("Unable to pull Docker image \""+containerStatus.getImage()+"\". Check if image name is spelled correctly");
+                        throw new IllegalStateException("BAD_DOCKER_IMAGE");
```
The exception is needed in the stack of `AllContainersRunningPodWatcher.areAllContainersRunning()`<- `periodicAwait()` <- `await()`.  

Creating a new exception type causes a clash in `AllContainersRunningPodWatcher.eventReceived()` which is overriding the parent method of `Watcher.eventReceived()`
To allow the aforementioned flow, it also causes the flow of `areAllContainersRunning()` <- `updateState()` <- `eventReceived()` 
Would you suggest:

1. catching the exception at `updateState()` or `eventReceived()` such that `eventReceived()` has to also pass it along.
2. Remain with the branching usage of the `IllegalStateException`, which `eventReceived()` already throws
3. Some alternative ?

Vlatombe replied:
Make your exception extend `RuntimeException`.

pyieh commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:
```diff
@@ -213,6 +239,17 @@ public void launch(SlaveComputer computer, TaskListener listener) {
+    /* itemTaskName is format of "part of <ORGANIZATION> <JOB NAME> >> <BRANCH> #<BUILD NUMBER> */
+    /* <ORGANIZATION> and <BRANCH> are only there if Pipeline created through BlueOcean, else just <JOB NAME>  */
+    private String getJobName(String itemTaskName) {
+        final String partOfStr = "part of ";
```
Removed this method as the only use was logging in the case there was an unexpected job name and not actually used in the feature functionality.

Vlatombe commented:
We have a flake (ContainerExecDecoratorTest.testCommandExecution) which can be ignored, but there are two other tests failing that need to be addressed.

```
kind / org.csanchez.jenkins.plugins.kubernetes.pipeline.KubernetesDeclarativeAgentTest.declarativeWithNonexistentDockerImage | 3 min 0 sec | 4
kind / org.csanchez.jenkins.plugins.kubernetes.pipeline.KubernetesDeclarativeAgentTest.declarativeWithNonexistentDockerImageLongLabel | 3 min 0 sec | 4
kind / org.csanchez.jenkins.plugins.kubernetes.pipeline.ContainerExecDecoratorTest.testCommandExecution
```

pyieh commented:
@Vlatombe I've fixed those tests. But the KubernetesPipelineTest.runInPodFromYaml is failing in Jenkins, but it's passing for me locally. 
Locally for me however, KubernetesPipelineTest.cascadeDelete is failing. Even a clean clone from the jenkinsci/kubernetes-plugin master is failing. 
Any suggestions?

pyieh commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:
```diff
@@ -124,12 +126,30 @@ public void launch(SlaveComputer computer, TaskListener listener) {
             try (Watch _w = client.pods().inNamespace(namespace1).withName(podName).watch(watcher)) {
                 watcher.await(template.getSlaveConnectTimeout(), TimeUnit.SECONDS);
+            } catch (InvalidDockerImageException e) {
+                if (e.getMessage().contains("Unable to pull Docker image")) {
```
Updated. Just an FYI this could become useful in the future if there are different error messages for different reasons why a Docker image can't be pulled. Have not seen this case yet though so removing for now.

pyieh commented:
@jglick please review. Thank you.
